### ISSUE TITLE
Multiclass Segmentation

### ISSUE THREAD

**Issue by polijn** [qubvel/segmentation_models.pytorch#798]
Hi,

I already used this framework for binary classification and it basically worked out of the box!
Thanks @qubvel  for this framework! 
Right now I want to use it for multiclass segmentation.
Im pretty much took the example from this git repository.

I changed the dice loss (thanks to all the issues in this repository) from  
``` python 
loss = smp.utils.losses.DiceLoss() # DEPREACHED, not suitable for multiclass?
```
 to 
``` python 
loss = smp.losses.DiceLoss(mode='multiclass')
``` 

because the first one was giving me the same loss value in every iteration.

with this new loss the first error message is: 
"one_hot is only applicable to index tensor."

like @qubvel mentioned before one should convert the target to long.

i modified my preprocessing in the data loader like this:
``` python
# apply preprocessing
if self.preprocessing:
    sample = self.preprocessing(image=image, mask=mask)
    image, mask = sample['image'], sample['mask']
    mask = torch.as_tensor(mask, dtype=torch.long) # Ensure mask is Long type for one_hot function 
    ## here mask has the dimension (5, 480, 480) 
```

now I'm getting this:
```bash
[/usr/local/lib/python3.10/dist-packages/segmentation_models_pytorch/losses/_functional.py](https://localhost:8080/#) in soft_dice_score(output, target, smooth, eps, dims)
    177     dims=None,
    178 ) -> torch.Tensor:
--> 179     assert output.size() == target.size()
    180     if dims is not None:
    181         intersection = torch.sum(output * target, dim=dims)
```

``` python
print(output.size())
>> torch.Size([10, 5, 230400]) # makes sense
print(target.size())
>> torch.Size([10, 5, 1152000]) # I don't understand
```

since debugging is very difficult in colab 
it would be amazing if anybody can help me.
as soon as I get it to work I will upload it, it seems that multiple people have struggled with multiclass and there are no examples out there.

**Comment by rongtongxueya** [comment]
same question. mark

**Comment by kevinsweeney84** [comment]
Did you get this worked out? Having same issue.

**Comment by ljb-1** [comment]
same question. mark

**Comment by polijn** [comment]
@kevinsweeney84 no due to time pressure, I needed to switch to a different repository.

**Comment by Dionysus326** [comment]
I've got it to work somehow.
Would share the solution later (because I am working on a pretty urgent project now
In short the problem is the format of label data.

**Comment by TalhaUsuf** [comment]
> I've got it to work somehow. Would share the solution later (because I am working on a pretty urgent project now In short the problem is the format of label data.

Can you please suggest solution?

**Comment by ChristianWiele** [comment]
I checked the source code. The issue is that mode="multiclass"  expects a non-one hot encoded target tensor, that is then one hot encoded. When feeding in a one hot encoded tensor, the two above issues occur: 1.) F.one_hot() requires a long tensor, and 2.) after one hot encoding the already one hot encoded tensor the shapes of preds and target tensor do not match anymore (causing the failed assertion).
There are two options. You can use a non-one hot encoded target tensor (as expected by the implementation, still requires a long tensor). Or you can use a one hot encoded target with mode="multilabel". But this only works if you apply an activation before calculating the loss (requires to set from_logits=False). 
Multilabel uses the same shape (N, C, H, W) as a one hot encoded target tensor. The difference is that the C-dimension can have more than a single 1 (multi-hot? encoding / each pixel can represent more than a single class). For each C-layer a binary classification is done. To avoid the latter you have to apply your own activation beforehand.

**Comment by Billy-ZTB** [comment]
> I checked the source code. The issue is that mode="multiclass" expects a non-one hot encoded target tensor, that is then one hot encoded. When feeding in a one hot encoded tensor, the two above issues occur: 1.) F.one_hot() requires a long tensor, and 2.) after one hot encoding the already one hot encoded tensor the shapes of preds and target tensor do not match anymore (causing the failed assertion). There are two options. You can use a non-one hot encoded target tensor (as expected by the implementation, still requires a long tensor). Or you can use a one hot encoded target with mode="multilabel". But this only works if you apply an activation before calculating the loss (requires to set from_logits=False). Multilabel uses the same shape (N, C, H, W) as a one hot encoded target tensor. The difference is that the C-dimension can have more than a single 1 (multi-hot? encoding / each pixel can represent more than a single class). For each C-layer a binary classification is done. To avoid the latter you have to apply your own activation beforehand.

Hello! I read your comment but I still cannot figure out how to solve this problem. What I wanted to do by this code is to classify the components in remote sensing images into 6 classes. My images are 256*256 RGB pictures(.tif) , while my label images are 256*256 images(.png) . I set mode="multilabel" and got the assertion error. I printed the size of predicted and y :
predicted:torch.Size([3, 6, 256, 256])     y: torch.Size([3, 3, 256, 256]).
output.size() and target.size() in soft_dice_score : torch.Size([3, 6, 65536])   target size  torch.Size([3, 6, 32768]). I noticed that 32768 happens to be half of 65536(=256*256), because the target channels 3 is half of the predicted channels 6. Thus caused the assertion error. I think maybe the output channels should be transformed to 3 to express RGB colors, would it be possible?

**Comment by Billy-ZTB** [comment]
> Hi,
> 
> I already used this framework for binary classification and it basically worked out of the box! Thanks @qubvel for this framework! Right now I want to use it for multiclass segmentation. Im pretty much took the example from this git repository.
> 
> I changed the dice loss (thanks to all the issues in this repository) from
> 
> ```python
> loss = smp.utils.losses.DiceLoss() # DEPREACHED, not suitable for multiclass?
> ```
> 
> to
> 
> ```python
> loss = smp.losses.DiceLoss(mode='multiclass')
> ```
> 
> because the first one was giving me the same loss value in every iteration.
> 
> with this new loss the first error message is: "one_hot is only applicable to index tensor."
> 
> like @qubvel mentioned before one should convert the target to long.
> 
> i modified my preprocessing in the data loader like this:
> 
> ```python
> # apply preprocessing
> if self.preprocessing:
>     sample = self.preprocessing(image=image, mask=mask)
>     image, mask = sample['image'], sample['mask']
>     mask = torch.as_tensor(mask, dtype=torch.long) # Ensure mask is Long type for one_hot function 
>     ## here mask has the dimension (5, 480, 480) 
> ```
> 
> now I'm getting this:
> 
> ```shell
> [/usr/local/lib/python3.10/dist-packages/segmentation_models_pytorch/losses/_functional.py](https://localhost:8080/#) in soft_dice_score(output, target, smooth, eps, dims)
>     177     dims=None,
>     178 ) -> torch.Tensor:
> --> 179     assert output.size() == target.size()
>     180     if dims is not None:
>     181         intersection = torch.sum(output * target, dim=dims)
> ```
> 
> ```python
> print(output.size())
> >> torch.Size([10, 5, 230400]) # makes sense
> print(target.size())
> >> torch.Size([10, 5, 1152000]) # I don't understand
> ```
> 
> since debugging is very difficult in colab it would be amazing if anybody can help me. as soon as I get it to work I will upload it, it seems that multiple people have struggled with multiclass and there are no examples out there.

Did you have the problem that when you set mode='multiclass' and batch_size isn't 1, the prediction size doesn't meet the ground truth size?
